waiter/test/waiter/descriptor_test.clj

```diff
@@ -1111,4 +1113,35 @@
                       :service-description {"permitted-user" "*"
                                             "run-as-user" auth-user}
                       :service-preauthorized false}]
-      (is (true? (service-invocation-authorized? can-run-as? auth-user descriptor))))))
\ No newline at end of file
+      (is (true? (service-invocation-authorized? can-run-as? auth-user descriptor))))))
+
+(deftest test-extract-service-state
+  (let [fn-name "test-extract-service-state"
+        router-id "r1"
+        service-id "s1"
+        retrieve-service-status-label-fn (constantly nil)]
+
+    (testing (str fn-name ":ping-result-received-response")
+      (let [ping-result :received-response
+            goal-state "healthy"
+            make-inter-router-requests-async-fn (fn [endpoint & _]
+                                                  (is (= endpoint (str "apps/" service-id "/await/" goal-state)))
+                                                  {"r1" (async/go {:body (async/go (json/write-str {"success?" true}))})
+                                                   "r2" (async/go {:body (async/go (json/write-str {"success?" true}))})})
+            fallback-state-atom (atom {:available-service-ids #{"s1"}
+                                       :healthy-service-ids #{"s1"}})
+            service-state (<?? (extract-service-state router-id retrieve-service-status-label-fn fallback-state-atom make-inter-router-requests-async-fn service-id ping-result))]
+        (is (= (:service-id service-state) service-id))
+        (is (true? (:exists? service-state)))
+        (is (true? (:healthy? service-state)))))
+
+    (testing (str fn-name ":ping-result-timed-out")
+      (let [ping-result :timed-out
+            make-inter-router-requests-async-fn (fn [& _]
+                                                  (throw (Exception. "Should not call this function")))
```

Comment:
Can we have a failing assertion instead of/in addition to throwing the exception?
That way the test will fail if some other piece of code is handling the exception.